src/coreclr/vm/codeman.cpp

```diff
@@ -1930,6 +1923,131 @@ void EEJitManager::SetCpuInfo()
+
+            if (familyId == 0x06)
+            {
+                if (extendedModelId == 0x05)
+                {
+                    if (modelId == 0x05)
+                    {
+                        // * Skylake (Server)
+                        // * Cascade Lake
+                        // * Cooper Lake
+
+                        if (preferredVectorBitWidth == 0)
+                        {
+                            preferredVectorBitWidth = 256;
+                        }
+                    }
+                }
+                else if (extendedModelId == 0x06)
+                {
+                    if (modelId == 0x06)
+                    {
+                        // * Cannon Lake
+
+                        if (preferredVectorBitWidth == 0)
+                        {
+                            preferredVectorBitWidth = 256;
+                        }
+                    }
+                }
+            }
+        }
+
+        if (maxVectorTBitWidth == 0)
+        {
+            // We currently default to a maximum Vector<T> of 256-bits on AVX512 hardware
+            maxVectorTBitWidth = 256;
+        }
+
+        if (preferredVectorBitWidth == 0)
+        {
+            // We currently default to a preferred vector width of 512 on AVX512 hardware
+            preferredVectorBitWidth = 512;
+        }
+
+        maxVectorBitWidth = 512;
+    }
+    else if (instructionSetFlags.HasInstructionSet(InstructionSet_AVX))
+    {
+        if (instructionSetFlags.HasInstructionSet(InstructionSet_AVX2))
+        {
+            if (maxVectorTBitWidth == 0)
+            {
+                // We currently default to a maximum Vector<T> of 256-bits on AVX2 hardware
+                maxVectorTBitWidth = 256;
+            }
+
+            if (preferredVectorBitWidth == 0)
+            {
+                // We currently default to a preferred vector width of 256-bits on AVX2 hardware
+                preferredVectorBitWidth = 256;
+            }
+        }
+
+        maxVectorBitWidth = 256;
+    }
+    else
+    {
+        maxVectorBitWidth = 128;
+    }
+#elif defined(TARGET_ARM64)
+    // This may need to change if SVE support is introduced
+    maxVectorBitWidth = 128;
+#else
+    // Vector<T> is always at least 128-bits in the fallback case
+    maxVectorBitWidth = 128;
+#endif
+
+    // clamp(maxVectorTBitWidth, minVectorBitWidth, maxVectorBitWidth)
+    maxVectorTBitWidth = max(maxVectorTBitWidth, minVectorBitWidth);
+    maxVectorTBitWidth = min(maxVectorTBitWidth, maxVectorBitWidth);
+
+    // clamp(preferredVectorBitWidth, minVectorBitWidth, maxVectorBitWidth)
+    preferredVectorBitWidth = max(preferredVectorBitWidth, minVectorBitWidth);
+    preferredVectorBitWidth = min(preferredVectorBitWidth, maxVectorBitWidth);
+
+    CORJIT_VECTOR_INFO jitVectorInfo;
+
+    jitVectorInfo.MaxVectorTBitWidth      = maxVectorTBitWidth;
+    jitVectorInfo.PreferredVectorBitWidth = preferredVectorBitWidth;
+
+    m_jitVectorInfo = jitVectorInfo;
```

Comment:
CC. @jkotas, this would benefit from your input on the general approach and VM changes.

As indicated in the top post, we've traditionally avoided doing any kind of "microarchitectural specific optimizations" and we've often been able to workaround such limitations fairly trivially without significant impact to performance.

However, with AVX-512 we're in a somewhat more unique situation where we have 2-4x or more perf improvements available, but not on the very first generation of AVX-512 capable hardware when some smaller workloads are available.

We also have the consideration that we have `Vector<T>` which users cannot currently "pick" the size of without disabling certain ISAs entirely and that means that it cannot grow to 512-bits for apps where it would be beneficial which limits our ability to version it moving forward (both for AVX-512 on x64 and eventually SVE on Arm64).

The "fix" here is to expose two new knobs. The first allows controlling the maximum size of `Vector<T>`, thus while we won't change the default from `256-bits` today (to avoid regressing existing code paths that may be handling small scenarios), users can opt-in to `512-bits` if working in big data scenarios where it benefits them -or- alternatively they can shrink to `128-bits` if that is beneficial instead.

Secondly there is a knob that allows controlling the maximum size used for any implicit vectorization done by the JIT (such as our `memset`, `memmov`, and `memcmp` unrolling). This defaults to the largest available since we switch vector size based on the input size already, but also means that the two micro-architectures that have heavy throttling for AVX-512 usage can opt-out by default and use 256-bits instead.

In both cases this does not disable the ability to use the higher ISAs explicitly (like `DOTNET_EnableAVX2=0` would), it just helps influence what the JIT does to ensure we don't accidentally hurt performance for the older hardware while still allowing the otherwise massive benefits that EVEX provides and that are seen on newer AVX512 capable hardware.

Reply:
The knob to configure maximum size of `Vector<T>` makes sense to me. I agree that we will need that to migrate Vector<T> to AVX-512 and SVE.

> a somewhat more unique situation where we have 2-4x or more perf improvements available, but not on the very first generation of AVX-512 capable hardware when some smaller workloads are available.

We had very similar discussion about the specific AMD hardware with poor implementation of some of the vector instructions. My take on these situations has been to detect the hardware with poor implementation of the specific instruction set and report the instruction set as not available by default if the hardware implements it poorly (allowing users to override it is fine). It makes things simple and the problem is going to disappear over time as the poor hardware gets old. (IIRC, we did not take my suggestion for the AMD case.)

I am not fan of the knob to configure maximum size of implicit vectorization used by the JIT. It exists just to compensate for hardware with poor AVX512 implementation. I think it would be better to instead report the AVX512 as not available on this poor hardware by default.

Reply:
In the previous case (with AMD) it was two very specific and largely edge cases instructions (`PDEP` and `PEXT`) which we were able to trivially work around. We also didn't want to be in a scenario where we disabled an ISA where the rest of the exposed APIs were fast and more commonly used.

This is a somewhat similar case. AVX512 is a new ISA that brings in 4 key pieces of functionality:
1. 512-bit vectors
2. Per instruction masking support
3. An extended register set (16 additional floating-point registers)
4. New instructions for 128-bit, 256-bit, and 512-bit scenarios

For the hardware in question, the only case that is pessimized is `1` and only for small workloads. For large workloads it remains fast/performant. The masking support, the extended registers, and the new instructions when using their 128-bit or 256-bit variants do not incur the downclocking penalty in question and so do not cause any downsides in their use (beyond the same considerations all SIMD code incurs).

We could disable AVX-512 support entirely for Skylake-X and Cooper Lake, but it's not my preferred option. I don't think the switch will become outdated over time either, I think rather we may see more prevalence in the future as hardware continues to become more heterogenous. That is, I expect the desire for such a configuration knob will increase as we see an increase in hardware that supports both big and little cores, as we see an increase in hardware functionality like SVE which can range from 128 to 2048 bits on a per chip basis, and as we see an increase in desire to run apps that respect energy efficiency settings.

Another option would be to keep it internal only for the time being. This would still allow JIT light-up for AVX512 features 2/3/4 on the hardware in question. Users just wouldn't be able to opt into 512-bit usage even for HPC scenarios. For user code that cares, it can always cache the same CPUID checks in a `static readonly bool` and do the same checks themselves.

Reply:
> In the previous case (with AMD) it was two very specific and largely edge cases instructions (PDEP and PEXT)

Nit: IIRC, the affected ISA had 4 instructions total, two of those were poorly implemented by the affected AMD hardware, and the other two were not very crucial.

> For the hardware in question, the only case that is pessimized is 1 and only for small workloads.

It may be only 1 case and only for small workloads, but it is bad enough that we want to disable it for implicit JIT vectorization by default. I think it is likely that every other library out there that tries to use AVX512 hits the same problem.

> I think rather we may see more prevalence in the future as hardware continues to become more heterogenous.

Nit: I doubt it. On client/mobile, the heterogenous hardware manifest itself as other types of power management. In cloud, the complexity involved in scheduling heterogenous hardware is unlikely to make it worth it.

> Another option would be to keep it internal only for the time being.

Sounds good me. I do not mind if we have undocumented JIT switches that motivated folks can play with, the same way as we have undocumented JIT switches for tweaking inliner behavior, etc. I think trying to document and explain how to configure well in official docs is a lot of work, and it does not add value at the end.

Reply:
> Nit: IIRC, the affected ISA had 4 instructions total, two of those were poorly implemented by the affected AMD hardware, and the other two were not very crucial.

We only publicly expose 4 of them (PDEP, PEXT, BZHI, MULX). We internally use the other 4 (RORX, SARX, SHRX, SHLX) via implicit lightup for most shift/rotate instructions as they are non-RMW and faster executing forms for the majority case, and thus are used fairly extensively in almost every application.

I've updated the switch to be `UNSUPPORTED_PreferredVectorBitWidth` so that it isn't one of the general documented cases.